tmc-langs-rust/src/main/java/fi/helsinki/cs/tmc/langs/rust/CargoResultParser.java

```diff
@@ -0,0 +1,121 @@
+import com.google.common.collect.ImmutableMap;
+
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Parses Cargos test output.
+ */
+public class CargoResultParser {
+
+    private static final RunResult PARSING_FAILED = new RunResult(
+            Status.GENERIC_ERROR,
+            ImmutableList.<TestResult>of(),
+            new ImmutableMap.Builder<String, byte[]>()
+            .put("generic_error_message", "Test results couldn't be parsed."
+                    .getBytes(StandardCharsets.UTF_8))
+            .build());
+
+    //test result: FAILED. 25 passed; 1 failed; 0 ignored; 0 measured
+    private static final Pattern RESULT =
+            Pattern.compile("test result: .*\\. (?<passes>\\d*) passed; (?<fails>\\d*) failed; \\d* ignored; \\d* measured");
+    //test test::dim4::test_4d_1_80_perioditic ... ok
+    private static final Pattern TEST = Pattern.compile("test (?<name>.*) \\.\\.\\. .*");
+    //thread 'test::dim2::test_2d_1_80_normal' panicked at 'assertion failed: false', src\test\dim2.rs:12
+    private static final Pattern FAILURES =
+            Pattern.compile(".*thread '(?<name>.*)' panicked at '(?<description>.*)', .*");
+
+    /**
+     * Parses given process results outputs to a run result.
+     */
+    public RunResult parse(ProcessResult processResult) {
+        String output = processResult.output;
+        String[] lines = output.split("\\r?\\n");
+
+        Matcher matcher = RESULT.matcher(output);
+        while (matcher.find()) {
+            int fails = Integer.parseInt(matcher.group("fails"));
+            int passes = Integer.parseInt(matcher.group("passes"));
+            if (fails + passes > 0) {
+                Optional<Map<String, String>> failures
+                        = findFailures(processResult.output, fails);
+                if (failures.isPresent()) {
+                    Status status = fails == 0 ? Status.PASSED : Status.TESTS_FAILED;
+                    Optional<List<String>> tests = parseResults(lines);
+                    if (tests.isPresent()) {
+                        return new RunResult(
+                                status,
+                                buildTestResults(tests.get(), failures.get()),
+                                new ImmutableMap.Builder()
+                                .put(SpecialLogs.STDOUT,
+                                        processResult.output.getBytes(StandardCharsets.UTF_8))
+                                .put(SpecialLogs.STDERR,
+                                        processResult.errorOutput.getBytes(StandardCharsets.UTF_8))
+                                .build());
+                    }
+                }
+                break;
+            }
+        }
+        return PARSING_FAILED;
+    }
+
+    private Optional<List<String>> parseResults(String[] lines) {
```

Comment:
Why `Optional` if we never return `Optional.absent()`: can't we just return the empty list?